include/unifex/async_invoke.hpp

```diff
@@ -17,49 +17,43 @@
 
 #include <unifex/config.hpp>
 #include <unifex/tag_invoke.hpp>
+#include <unifex/type_traits.hpp>
 
 #include <functional>
 
 #include <unifex/detail/prologue.hpp>
 
 namespace unifex {
 
-namespace _tag {
-template <typename...>
-struct tag;
-} // namespace _tag
-template <typename... Ts>
-using tag = _tag::tag<Ts...>*;
-
-namespace _co_invoke {
+namespace _async_invoke {
 inline constexpr struct _fn {
     template (typename Fn, typename... Args)
       (requires tag_invocable<
           _fn,
-          tag<std::invoke_result_t<Fn, Args...>, Fn, Args...>,
+          type_identity<std::invoke_result_t<Fn, Args...>>,
           Fn,
           Args...>)
     UNIFEX_ALWAYS_INLINE constexpr auto operator()(Fn&& fn, Args&&... args) const
       noexcept(is_nothrow_tag_invocable_v<
           _fn,
-          tag<std::invoke_result_t<Fn, Args...>, Fn, Args...>,
+          type_identity<std::invoke_result_t<Fn, Args...>>,
           Fn,
           Args...>)
       -> tag_invoke_result_t<
           _fn,
-          tag<std::invoke_result_t<Fn, Args...>, Fn, Args...>,
+          type_identity<std::invoke_result_t<Fn, Args...>>,
           Fn,
           Args...> {
       return tag_invoke(
           *this,
-          tag<std::invoke_result_t<Fn, Args...>, Fn, Args...>{},
+          type_identity<std::invoke_result_t<Fn, Args...>>{},
           (Fn&&) fn,
           (Args&&) args...);
     }
-} co_invoke{};
+} async_invoke{};
```

Comment:
It's not necessarily "async", though is it?
I can imagine we might customise this for `generator` to also allow capturing args by value?

Maybe `copy_invoke` or `by_value_invoke`?
Or `no_dangling_refs_invoke`? ;)

Reply:
Is this just a flavor of `let`?

Reply:
Yes, I think you can kind of do the same thing with `let`.

```
co_invoke(f, args...)
```
is roughly equivalent to
```
let(just(args...), f)
```